File: src/chipschallenge/blockreactions/RedLockTo.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.Inventory.Key;

public class RedLockTo extends NoSlipReaction {

    private RedLockTo() {
    }
    private static RedLockTo mInstance = null;

    public static synchronized RedLockTo getInstance() {
        if (mInstance == null) {
            mInstance = new RedLockTo();
        }
        return mInstance;
    }

    public void react(Block moving, Block standing) {
        useKey(Key.RED);
        standing.destroy();
    }

    public boolean canMove(Block moving, Block standing) {
        return moving.isChip() && hasKey(Key.RED);
    }
}

File: ChipsChallenge.Shared/Blockreactions/RedLockTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    using Key = Inventory.Key;

    public class RedLockTo : NoSlipReaction
    {
        private static RedLockTo instance;

        private RedLockTo()
        {
        }

        public static RedLockTo Instance
        {
            get
            {
                lock (typeof(RedLockTo))
                {
                    return instance ?? (instance = new RedLockTo());
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
            UseKey(Key.RED);
            standing.Destroy();
        }

        public override bool canMove(Block moving, Block standing)
        {
            return moving.Chip && HasKey(Key.RED);
        }
    }
}